pkg/controllers/disruption/controller.go

```diff
@@ -260,31 +261,36 @@ func (c *Controller) createReplacementNodeClaims(ctx context.Context, m Method,
 	return nodeClaimNames, nil
 }
 
-func (c *Controller) MarkDisrupted(ctx context.Context, m Method, candidates ...*Candidate) error {
-	stateNodes := lo.Map(candidates, func(c *Candidate, _ int) *state.StateNode {
-		return c.StateNode
-	})
-	if err := state.RequireNoScheduleTaint(ctx, c.kubeClient, true, stateNodes...); err != nil {
-		return serrors.Wrap(fmt.Errorf("tainting nodes, %w", err), "taint", pretty.Taint(v1.DisruptedNoScheduleTaint))
-	}
-
-	providerIDs := lo.Map(candidates, func(c *Candidate, _ int) string { return c.ProviderID() })
-	c.cluster.MarkForDeletion(providerIDs...)
-
+func (c *Controller) MarkDisrupted(ctx context.Context, m Method, candidates ...*Candidate) ([]*Candidate, error) {
 	errs := make([]error, len(candidates))
 	workqueue.ParallelizeUntil(ctx, len(candidates), len(candidates), func(i int) {
+		if err := state.RequireNoScheduleTaint(ctx, c.kubeClient, true, candidates[i].StateNode); err != nil {
+			errs[i] = serrors.Wrap(fmt.Errorf("tainting nodes, %w", err), "taint", pretty.Taint(v1.DisruptedNoScheduleTaint))
+			return
+		}
 		// refresh nodeclaim before updating status
 		nodeClaim := &v1.NodeClaim{}
-
-		if err := c.kubeClient.Get(ctx, client.ObjectKeyFromObject(candidates[i].NodeClaim), nodeClaim); err != nil {
+		if err := retry.OnError(retry.DefaultBackoff, func(err error) bool { return client.IgnoreNotFound(err) != nil }, func() error {
```

Comment:
Is there a specific set of errors we want to retry for here? The only error I think I'd expect us to see which wouldn't be transient is an RBAC permission issue. I don't think we need to consider that case since Karpenter needs RBAC permissions to read / write to nodes to operate but I'm wondering if there are any other cases we wouldn't want to retry on.

Reply:
> Yeah, I think if you start talking about RBAC permissions, things just start failing generally -- I'd rather not handle it here

Yeah, I agree. I haven't been able to think of anything else other than RBAC permissions, so this seems good to me.